source/worksheet/worksheet.cpp

```diff
@@ -736,6 +740,128 @@ void worksheet::clear_row(row_t row)
     // TODO: garbage collect newly unreferenced resources such as styles?
 }
 
+void worksheet::insert_rows(row_t row, std::uint32_t amount)
+{
+    move_cells(row, amount, row_or_col_t::row);
+}
+
+void worksheet::insert_columns(column_t column, std::uint32_t amount)
+{
+    move_cells(column.index, amount, row_or_col_t::column);
+}
+
+void worksheet::delete_rows(row_t row, std::uint32_t amount)
+{
+    move_cells(row + amount, amount, row_or_col_t::row, true);
+}
+
+void worksheet::delete_columns(column_t column, std::uint32_t amount)
+{
+    move_cells(column.index + amount, amount, row_or_col_t::column, true);
+}
+
+void worksheet::move_cells(std::uint32_t min_index, std::uint32_t amount, row_or_col_t row_or_col, bool reverse)
+{
+    std::vector<detail::cell_impl> tmp_cell_map;
+    std::unordered_map<row_t, xlnt::row_properties> tmp_row_properties;
+    std::unordered_map<column_t, xlnt::column_properties> tmp_column_properties;
+
+    auto cell_iter = d_->cell_map_.cbegin();
+
+    std::uint32_t current_index = 0;
+
+    while (cell_iter != d_->cell_map_.cend())
+    {
+        switch (row_or_col)
+        {
+        case row_or_col_t::row:
+            current_index = cell_iter->first.row();
+            break;
+        case row_or_col_t::column:
+            current_index = cell_iter->first.column().index;
+            break;
+        default:
+            throw xlnt::unhandled_switch_case();
+        }
+
+        // Clear destination cells when going in reverse
+        if (reverse && current_index < min_index - amount)
```

Comment:
Could you care to elaborate on that?

Reply:
The reason why amount is added to the `min_index` when calling `move_cells` from the delete methods is so that it translates 'delete rows 2,3,4 to move cells from rows 5+ up 3 rows'. The additional step of deleting cells when reverse=true is to guarantee that those cells will be overwritten, something not required when moving forward (down/right). This, besides the sign of the amount is the only difference between inserting and deleting.

Reply:
I was thinking it would make the "these cells can be ignored" condition easier
```
if (reverse) { min_index -= amount; }
if (index < min_index) continue; // nothing to be done
if (reverse && index < (min_index + amount)) {} // delete
else {} // save the cells to move to new locations
```

Reply:
Ah I see, although taking the property stuff out of the way leaves just that.. whether you subtract first or not doesn't really change much..